lambda_functions/va_profile/va_profile_opt_in_out_lambda.py

```diff
@@ -448,6 +493,157 @@ def make_PUT_request(
+        )
+
+    except Exception as e:
+        logger.exception(
+            'An error occurred while attempting to send Comp and Pen opt-in confirmation SMS notification to vaProfileId %s: %s',
+            va_profile_id,
+            e,
+        )
+    return None
+
+
+def save_notification_id_to_cache(va_profile_id: int, notification_id: str, source_date: str):
+    """
+    Update the VAProfileLocalCache table by inserting the notification_id for the given va_profile_id.
+
+    Args:
+        va_profile_id (int): The VA profile ID of the user the notification was sent to.
+        notification_id (str): The notification UUID.
+    """
+    try:
+        with db_connection.cursor() as cursor:
+            # Execute the SQL query with the provided parameters.
+            cursor.execute(OPT_IN_OUT_ADD_NOTIFICATION_ID_QUERY, (notification_id, va_profile_id, source_date))
+
+            db_connection.commit()
+            logger.info(
+                'Successfully updated VAProfileLocalCache with notification_id %s for va_profile_id %s.',
+                notification_id,
+                va_profile_id,
+            )
+
+    except psycopg2.IntegrityError as e:
+        db_connection.rollback()
+        logger.error(
+            'Failed to insert notification_id %s for va_profile_id %s. IntegrityError: %s',
+            notification_id,
+            va_profile_id,
+            str(e),
+        )
+    except Exception as e:
+        db_connection.rollback()
+        logger.error(
+            'An error occurred while attempting to update notification_id for va_profile_id %s: %s',
+            va_profile_id,
+            str(e),
+        )
+    finally:
+        cursor.close()
+
+
+def base64url(source: bytes) -> str:
+    """
+    Encode a byte source to a base64 URL-safe string.
+
+    Args:
+        source (bytes): The byte source to be encoded.
+
+    Returns:
+        str: The base64 URL-safe encoded string.
+    """
+    encoded_source = base64.b64encode(source).decode('utf-8')
+    encoded_source = encoded_source.rstrip('=').replace('+', '-').replace('/', '_')
+    return encoded_source
+
+
+def generate_jwt(service_api_key: str, service_id: str) -> str:
+    """
+    Generate a JWT token for authentication purposes.
+
+    Args:
+        service_api_key (str): The API key used to sign the JWT.
+        service_id (str): The service identifier.
+
+    Returns:
+        str: The generated JWT token.
+    """
+    header = {'typ': 'JWT', 'alg': 'HS256'}
+    current_timestamp = int(time.time())
+
+    data = {'iss': service_id, 'iat': current_timestamp}
+
+    encoded_header = base64url(json.dumps(header).encode('utf-8'))
+    encoded_data = base64url(json.dumps(data).encode('utf-8'))
+
+    token = '{}.{}'.format(encoded_header, encoded_data)
+
+    signature = hmac.new(service_api_key.encode('utf-8'), token.encode('utf-8'), hashlib.sha256).digest()
+    encoded_signature = base64url(signature)
+
```

Comment:
the pyJWT library takes care of all of this.

Reply:
example:
```
import jwt
import os
import datetime

# example, set these as needed
# os.environ["SERVICE_API_KEY"] = "your-secret-key"
# os.environ["SERVICE_ID"] = "your-service-id"

# Retrieve secret and service ID from environment variables
jwt_secret = os.getenv("SERVICE_API_KEY")
service_id = os.getenv("SERVICE_ID")

headers = {
    "typ": "JWT",
    "alg": "HS256"
}

current_timestamp = int(datetime.datetime.now().timestamp())
payload = {
    "iss": service_id,
    "iat": current_timestamp
}

# Generate and return the signed JWT token using the pyJWT library
return jwt.encode(payload, jwt_secret, algorithm="HS256", headers=headers)

```
obviously, modify the vars to be what we need. no need for hmac or base64 encoding or two methods, pyJWT is a one stop shop.

Reply:
Thank you for providing this.

Reply:
Updated.